File: app/src/main/java/com/bis5/fitjourney/fragments/WorkoutDetailFragment.java
package com.bis5.fitjourney.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import com.bis5.fitjourney.R;
import com.bis5.fitjourney.adapters.ExerciseAdapter;
import com.bis5.fitjourney.adapters.WorkoutHistoryAdapter;
import com.bis5.fitjourney.databinding.FragmentWorkoutDetailBinding;
import com.bis5.fitjourney.models.AppDatabase;
import com.bis5.fitjourney.models.WorkoutLog;
import com.bis5.fitjourney.viewmodels.WorkoutViewModel;
import com.bis5.fitjourney.viewmodels.WorkoutViewModelFactory;

public class WorkoutDetailFragment extends Fragment implements WorkoutHistoryAdapter.OnLogClickListener {

    private FragmentWorkoutDetailBinding binding;
    private WorkoutViewModel workoutViewModel;
    private ExerciseAdapter exerciseAdapter;
    private WorkoutHistoryAdapter historyAdapter;
    private long workoutId;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentWorkoutDetailBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        if (getArguments() != null) {
            workoutId = WorkoutDetailFragmentArgs.fromBundle(getArguments()).getWorkoutId();
        }

        AppDatabase database = AppDatabase.getDatabase(requireContext());
        WorkoutViewModelFactory viewModelFactory = new WorkoutViewModelFactory(database.workoutDao(), database.exerciseDao(), database.workoutLogDao(), database.setLogDao());
        workoutViewModel = new ViewModelProvider(this, viewModelFactory).get(WorkoutViewModel.class);

        setupRecyclerViews();
        observeViewModel();
        setupClickListeners();
    }

    private void observeViewModel() {
        workoutViewModel.getWorkout(workoutId).observe(getViewLifecycleOwner(), workout -> {
            if (workout != null && getActivity() instanceof AppCompatActivity) {
                ActionBar actionBar = ((AppCompatActivity) getActivity()).getSupportActionBar();
                if (actionBar != null) {
                    actionBar.setTitle(workout.getName());
                }
            }
        });

        workoutViewModel.getExercises(workoutId).observe(getViewLifecycleOwner(), exercises -> {
            if(exercises != null) exerciseAdapter.updateExercises(exercises);
        });

        workoutViewModel.getWorkoutHistory(workoutId).observe(getViewLifecycleOwner(), history -> {
            if(history != null) historyAdapter.updateLogs(history);
        });
    }

    private void setupClickListeners() {
        binding.btnAddExercise.setOnClickListener(v -> showAddExerciseDialog());

        binding.btnStartWorkout.setOnClickListener(button -> {
            button.setEnabled(false); // Prevent double clicks
            workoutViewModel.startWorkout(workoutId).observe(getViewLifecycleOwner(), workoutLogId -> {
                if (workoutLogId != null) {
                    NavController navController = NavHostFragment.findNavController(this);
                    WorkoutDetailFragmentDirections.ActionWorkoutDetailFragmentToActiveWorkoutFragment action =
                            WorkoutDetailFragmentDirections.actionWorkoutDetailFragmentToActiveWorkoutFragment(workoutId, workoutLogId);
                    navController.navigate(action);
                }
            });
        });
    }

    @Override
    public void onResume() {
        super.onResume();
        if (binding != null) {
            binding.btnStartWorkout.setEnabled(true);
        }
    }

    private void setupRecyclerViews() {
        exerciseAdapter = new ExerciseAdapter();
        binding.rvExercises.setLayoutManager(new LinearLayoutManager(getContext()));
        binding.rvExercises.setAdapter(exerciseAdapter);

        historyAdapter = new WorkoutHistoryAdapter(this);
        binding.rvHistory.setLayoutManager(new LinearLayoutManager(getContext()));
        binding.rvHistory.setAdapter(historyAdapter);
    }

    private void showAddExerciseDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(requireContext());
        LayoutInflater inflater = requireActivity().getLayoutInflater();
        View dialogView = inflater.inflate(R.layout.dialog_add_exercise, null);

        final EditText etExerciseName = dialogView.findViewById(R.id.etExerciseName);
        final EditText etSets = dialogView.findViewById(R.id.etSets);
        final EditText etReps = dialogView.findViewById(R.id.etReps);
        final EditText etWeight = dialogView.findViewById(R.id.etWeight);

        builder.setView(dialogView)
                .setTitle("Add New Exercise")
                .setPositiveButton("Add", (dialog, id) -> {
                    String name = etExerciseName.getText().toString().trim();
                    String setsStr = etSets.getText().toString();
                    String repsStr = etReps.getText().toString();
                    String weightStr = etWeight.getText().toString();

                    if (!name.isEmpty()) {
                        int sets = setsStr.isEmpty() ? 0 : Integer.parseInt(setsStr);
                        int reps = repsStr.isEmpty() ? 0 : Integer.parseInt(repsStr);
                        double weight = weightStr.isEmpty() ? 0.0 : Double.parseDouble(weightStr);
                        //workoutViewModel.addExercise(workoutId, name, sets, reps, weight);
                    }
                })
                .setNegativeButton("Cancel", (dialog, id) -> dialog.cancel());

        builder.create().show();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    @Override
    public void onLogClick(WorkoutLog workoutLog) {
        NavController navController = NavHostFragment.findNavController(this);
        if(workoutLog != null) {
            //WorkoutDetailFragmentDirections.ActionWorkoutDetailFragmentToWorkoutSummaryFragment action = 
                //WorkoutDetailFragmentDirections.actionWorkoutDetailFragmentToWorkoutSummaryFragment(workoutLog.getId());
            //navController.navigate(action);
        }
    }
}

File: app/src/main/java/fragments/WorkoutDetailFragment.kt
package com.bis5.fitjourney.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.fragment.navArgs
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bis5.fitjourney.R
import com.bis5.fitjourney.databinding.FragmentWorkoutDetailBinding
import com.bis5.fitjourney.models.AppDatabase
import com.bis5.fitjourney.models.Exercise
import com.bis5.fitjourney.viewmodels.WorkoutViewModel
import com.bis5.fitjourney.viewmodels.WorkoutViewModelFactory

class WorkoutDetailFragment : Fragment() {

    private var _binding: FragmentWorkoutDetailBinding? = null
    private val binding get() = _binding!!

    private lateinit var workoutViewModel: WorkoutViewModel
    private lateinit var exerciseAdapter: ExerciseAdapter
    private val args: WorkoutDetailFragmentArgs by navArgs()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentWorkoutDetailBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val database = AppDatabase.getDatabase(requireContext())
        val viewModelFactory = WorkoutViewModelFactory(database.workoutDao(), database.exerciseDao())
        workoutViewModel = ViewModelProvider(this, viewModelFactory)[WorkoutViewModel::class.java]

        setupRecyclerView()

        workoutViewModel.getWorkout(args.workoutId).observe(viewLifecycleOwner) { workout ->
            binding.tvWorkoutName.text = workout?.name ?: "Workout Details"
        }

        workoutViewModel.getExercises(args.workoutId).observe(viewLifecycleOwner) { exercises ->
            exerciseAdapter.updateExercises(exercises)
        }

        binding.btnAddExercise.setOnClickListener {
            showAddExerciseDialog()
        }
    }

    private fun setupRecyclerView() {
        exerciseAdapter = ExerciseAdapter()
        binding.rvExercises.apply {
            layoutManager = LinearLayoutManager(context)
            adapter = exerciseAdapter
        }
    }

    private fun showAddExerciseDialog() {
        val context = requireContext()
        val dialogView = LayoutInflater.from(context).inflate(R.layout.dialog_add_exercise, null)
        val etExerciseName = dialogView.findViewById<EditText>(R.id.etExerciseName)
        val etSets = dialogView.findViewById<EditText>(R.id.etSets)
        val etReps = dialogView.findViewById<EditText>(R.id.etReps)
        val etWeight = dialogView.findViewById<EditText>(R.id.etWeight)

        AlertDialog.Builder(context)
            .setTitle("Add New Exercise")
            .setView(dialogView)
            .setPositiveButton("Add") { _, _ ->
                val name = etExerciseName.text.toString().trim()
                val sets = etSets.text.toString().toIntOrNull() ?: 0
                val reps = etReps.text.toString().toIntOrNull() ?: 0
                val weight = etWeight.text.toString().toDoubleOrNull() ?: 0.0

                if (name.isNotEmpty()) {
                    workoutViewModel.addExercise(args.workoutId, name, sets, reps, weight)
                }
            }
            .setNegativeButton("Cancel", null)
            .show()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}

class ExerciseAdapter(
    private var exercises: List<Exercise> = emptyList()
) : RecyclerView.Adapter<ExerciseAdapter.ExerciseViewHolder>() {

    class ExerciseViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val exerciseName: TextView = itemView.findViewById(R.id.tvExerciseName)
        val sets: TextView = itemView.findViewById(R.id.tvSets)
        val reps: TextView = itemView.findViewById(R.id.tvReps)
        val weight: TextView = itemView.findViewById(R.id.tvWeight)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ExerciseViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_exercise, parent, false)
        return ExerciseViewHolder(view)
    }

    override fun onBindViewHolder(holder: ExerciseViewHolder, position: Int) {
        val exercise = exercises[position]
        holder.exerciseName.text = exercise.name
        holder.sets.text = "Sets: ${exercise.sets}"
        holder.reps.text = "Reps: ${exercise.reps}"
        holder.weight.text = "Weight: ${exercise.weight} kg"
    }

    override fun getItemCount() = exercises.size

    fun updateExercises(newExercises: List<Exercise>) {
        exercises = newExercises
        notifyDataSetChanged()
    }
}